projects/vdk-plugins/vdk-jupyter/src/vdk/plugin/jupyter/notebook_based_step.py

```diff
@@ -0,0 +1,150 @@
+            x for x in directory.iterdir() if (x.name.lower().endswith(".ipynb"))
+        ]
+        script_files.sort(key=lambda x: x.name)
+        log.debug(f"Script files of {directory} are {script_files}")
+        return script_files
+
+
+class NotebookStepFuncFactory:
+    """
+    Implementations of runner_func for running Notebook steps
+    """
+
+    @staticmethod
+    def run_sql_step(step: NotebookStep, job_input: IJobInput) -> bool:
+        job_input.execute_query(step.code)
+        return True
+
+    @staticmethod
+    def run_python_step(step: NotebookStep, job_input: IJobInput) -> bool:
+        try:
+            sys.path.insert(0, str(step.job_dir))
+            filename = step.file_path.name
+            success = False
+            try:
+                log.debug("Loading %s ..." % filename)
+                module = types.ModuleType(step.name)
+                exec(step.code, module.__dict__)
+                log.debug("Loading %s SUCCESS" % filename)
+            except BaseException as e:
+                log.info("Loading %s FAILURE" % filename)
+                errors.log_and_rethrow(
+                    to_be_fixed_by=errors.ResolvableBy.USER_ERROR,
+                    log=log,
+                    what_happened="Failed loading job sources of %s" % filename,
+                    why_it_happened=errors.MSG_WHY_FROM_EXCEPTION(e),
+                    consequences=errors.MSG_CONSEQUENCE_TERMINATING_APP,
+                    countermeasures=errors.MSG_COUNTERMEASURE_FIX_PARENT_EXCEPTION
+                    + " Most likely importing a dependency or data job step failed, see"
+                    + " logs for details and fix the failed step (details in stacktrace).",
+                    exception=e,
+                    wrap_in_vdk_error=True,
+                )
+
+            for _, func in inspect.getmembers(module, inspect.isfunction):
+                if func.__name__ == "run":
+                    try:
+                        log.info("Entering %s#run(...) ..." % filename)
+                        NotebookStepFuncFactory.invoke_run_function(
+                            func, job_input, step.name
+                        )
+                        success = True
+                        return True
+                    finally:
+                        if success:
+                            log.info("Exiting  %s#run(...) SUCCESS" % filename)
+                        else:
+                            log.error("Exiting  %s#run(...) FAILURE" % filename)
+            log.warn(
+                "File %s does not contain a valid run() method. Nothing to execute. Skipping %s,"
+                + " and continuing with other files (if present).",
+                filename,
+                filename,
+            )
+            return success
+        finally:
+            sys.path.remove(str(step.job_dir))
+
+    @staticmethod
+    def invoke_run_function(func: Callable, job_input: IJobInput, step_name: str):
```

Comment:
We discussed face to face to add a TODO to consider removing duplication with vdk-core in the future. Now is not the correct moment to do so - https://refactoring.guru/refactoring/when